submissions/xT0PERx/Nanodegree-arcade-game/app.js

```diff
@@ -0,0 +1,102 @@
+const X_START_POSITION_PLAYER = 200;
+const Y_START_POSITION_PLAYER = 395;
+const X_START_POSITION_ENEMY = 0;
+const Y_START_POSITION_ENEMY = [55, 140, 225];
+const SPEED_ENEMY = [100, 200, 300];
+const STEP_LEFT_RIGHT = 100;
+const STEP_UP_DOWN = 85;
+const MAX_STEP_PLAYER = 400;
+const MIN_LENGTH_FIELD = 0;
+const MAX_LENGTH_FIELD = 500;
+const HALF_LENGTH_ENEMY = 50;
+
+let countWin = 0;
+// Enemies our player must avoid
+var Enemy = function (x, y, speed, objPlayer) {
+  this.x = x;
+  this.y = y;
+  this.speed = speed;
+  this.objPlayer = objPlayer;
+  this.sprite = "images/enemy-bug.png";
+};
+// Update the enemy's position, required method for game
+// Parameter: dt, a time delta between ticks
+Enemy.prototype.update = function (dt) {
+  this.x += this.speed * dt;
+  if (this.x > MAX_LENGTH_FIELD) {
+    this.x = -STEP_LEFT_RIGHT;
+    this.speed = Math.random() * MAX_LENGTH_FIELD + HALF_LENGTH_ENEMY;
+  }
+
+  this.collide();
+};
+// Draw the enemy on the screen, required method for game
+Enemy.prototype.render = function () {
+  ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+
+Enemy.prototype.collide = function () {
+  if (
+    this.objPlayer.y === this.y &&
+    this.objPlayer.x < this.x + HALF_LENGTH_ENEMY &&
+    this.objPlayer.x > this.x - HALF_LENGTH_ENEMY
+  ) {
+    this.objPlayer.goStartPosition();
+  }
+};
+const Player = function (x, y) {
+  this.x = x;
+  this.y = y;
+  this.sprite = "images/char-boy.png";
+};
+Player.prototype.update = function (dt) {};
+
+Player.prototype.render = function () {
+  ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+
+Player.prototype.handleInput = function (arr) {
+  if (arr === "left" && this.x > MIN_LENGTH_FIELD) {
+    this.x -= STEP_LEFT_RIGHT;
+  } else if (arr === "right" && this.x < MAX_STEP_PLAYER) {
+    this.x += STEP_LEFT_RIGHT;
+  } else if (arr === "up") {
+    this.y -= STEP_UP_DOWN;
+    if (this.y < MIN_LENGTH_FIELD) {
+      player.goStartPosition();
+      alert("Wins: " + ++countWin);
+    }
+  } else if (arr === "down" && this.y < Y_START_POSITION_PLAYER) {
+    this.y += STEP_UP_DOWN;
+  }
+};
+
+Player.prototype.goStartPosition = function () {
+  this.x = X_START_POSITION_PLAYER;
+  this.y = Y_START_POSITION_PLAYER;
+};
+
+const player = new Player(X_START_POSITION_PLAYER, Y_START_POSITION_PLAYER);
+
+const countEmemies = 3;
+const allEnemies = Array(countEmemies)
+  .fill()
+  .map(function (enemy, index) {
+    return new Enemy(
+      X_START_POSITION_ENEMY,
+      Y_START_POSITION_ENEMY[index],
+      SPEED_ENEMY[index],
+      player
+    );
+  });
```

Comment:
If you had an array of enemies settings like e.g.
```
const ENEMIES_INITIAL_SETTINGS = [{
  start_y_position: 100500,
  speed: 100500,
  }, {
  START_Y_POSITION: 100500,
  SPEED: 100500,
  } // etc
];
```
this fragment might have looked more elegant, e.g.
```
const enemies = 
  ENEMIES_INITIAL_SETTINGS.map(({start_y_position, speed}) => 
    new Enemy(ENEMY_START_X_POSITIOn, start_y_position, speed, .....)
  );
```

It's a bit unobvious to use `Array(x).fill()...` to replace iteration over the list of same-rank values.

Reply:
I agree, my optimization is somewhat more difficult to understand than yours. But I'm learning - at that moment I managed to come up with what I could. Should I remake this piece of code after your example?